executor/common_linux.h

```diff
@@ -2741,25 +2754,38 @@ static void initialize_vhci()
 	close(vhci_fd);
 	vhci_fd = kVhciFd;
 
-	struct vhci_vendor_pkt vendor_pkt;
-	if (read(vhci_fd, &vendor_pkt, sizeof(vendor_pkt)) != sizeof(vendor_pkt))
-		fail("read failed");
+	struct vhci_vendor_pkt_request vendor_pkt_req = {HCI_VENDOR_PKT, HCI_PRIMARY};
+	if (write(vhci_fd, &vendor_pkt_req, sizeof(vendor_pkt_req)) != sizeof(vendor_pkt_req))
+		fail("vendor_pkt_req write failed");
+
+	struct vhci_pkt vhci_pkt;
+	if (read(vhci_fd, &vhci_pkt, sizeof(vhci_pkt)) != sizeof(vhci_pkt))
+		fail("vhci_pkt read failed");
+
+	if (vhci_pkt.type == HCI_COMMAND_PKT && vhci_pkt.command_hdr.opcode == HCI_OP_RESET) {
+		char response[1] = {0};
```

Comment:
It is still possible that `HCI_VENDOR_PKT` comes before `HCI_COMMAND_PKT`, right? If yes, should we still read and handle the `HCI_COMMAND_PKT` afterwards?

Reply:
Yes, several `HCI_COMMAND_PKT` packets are coming and they are all handled by the `event_thread` (thread is created immediately after these lines). My current understanding is that the only exception is `HCI_OP_RESET` which can come before the `HCI_VENDOR_PKT`, so I handled that case separately here.

Reply:
If `HCI_VENDOR_PKT` comes before `HCI_OP_RESET`, I still don't see the code that would handle `HCI_OP_RESET` then. `event_thread` does not mention `HCI_OP_RESET`.

Reply:
Yeah, it's a bit tricky, `event_thread` calls `process_command_pkt` in case of `HCI_COMMAND_PKT` which [sends a dummy response](https://github.com/google/syzkaller/blob/ea3cf4e1819cfdbe693e0a6a2dea1a4082457beb/executor/common_linux.h#L2707) for every packet it does not understand including `HCI_OP_RESET`.

The BT stack works in a way that if you send more response bytes than it expects, it just ignores the remaining, so although the `dummy` response is `0xf9` bytes, and `HCI_OP_RESET` expects only 1 byte response, it will be still valid.

Also the [first byte of the command response](https://github.com/torvalds/linux/blob/0326074ff4652329f2a1a9c8685104576bd8d131/net/bluetooth/hci_sync.c#L260) packet contains the 1-byte `status / error code` field, where `0` means `SUCCESS`, which means the initialization process will continue. So sending the `dummy` response for most commands is satisfying (and somewhat future-proof in the case of the kernel starts sending new non-yet-implemented command request).